src/main/java/org/lwes/MemoryPool.java

```diff
@@ -0,0 +1,97 @@
+package org.lwes;
+
+import static org.lwes.Event.MAX_MESSAGE_SIZE;
+
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * this object maintains a memory pool for
+ *
+ */
+public class MemoryPool {
+
+  // Buffer pool is a queue of <Buffer>
+  static protected ConcurrentLinkedQueue<Buffer> bufferPool = new ConcurrentLinkedQueue<Buffer>();
+
+  /**
+   * Buffer object to keep a ByteBuffer and CharBuffer object
+   */
+  public static class Buffer {
+    private boolean    inPool = false;      // used to make sure one buffer is not
+                                            // put in to the pool more than once
+    private ByteBuffer byteBuffer;
+    private CharBuffer charBuffer;
+
+    /**
+     * @return the {@code CharBuffer} object of the buffer used as the input
+     *         for encoder
+     */
+    public CharBuffer getEncoderInputBuffer() {
+      return charBuffer;
+    }
+    /**
+     * @return the {@code ByteBuffer} object of the Buffer used as the output
+     *         for encoder
+     */
+    public ByteBuffer getEncoderOutputBuffer() {
```

Comment:
my own preference was to: as opposed to having an API that provides direct access to private attributes (because what is the point of having the attributes as private then?), only have APIs that does some operation on these internal attributes, e.g.,

public byte[] getEncoderOutputBufferBytes() { return byteBuffer.array(); }
public int getEncoderOutputBufferLength() { return byteBuffer.position(); }

but for this special case, Bytebuffer wraps a byte[] and there is really not much difference between the two approaches, but I am still perfectly Ok with changing this to the above if you guys prefer it that way.